cpp/src/fil/common.cuh

```diff
@@ -173,6 +172,89 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+namespace dispatch {
+
+template <template <bool, leaf_algo_t, int> class Func, typename storage_type,
```

Comment:
Isn't it possible wrap all other template parameters info `Func`? In this way, `Func` would accept only `N_ITEMS` as a template parameter, and the code would become more readable.

Reply:
We can do something like below. It adds 5-7 lines to every new `Func`, but might make the dispatch mechanism easier to read?
```C++
void compute_smem_footprint(predict_params& params) {
  return template <bool COLS_IN_SHMEM>
  [&]() {
    return template <int N_ITEMS>
    [&]() {
      return template <leaf_algo_t LEAF_ALGO>
      [&]() {
        // need GROVE_PER_CLASS_*_CLASSES
        if constexpr (LEAF_ALGO != GROVE_PER_CLASS) {
          ssp.shm_sz = ssp.get_smem_footprint<n_items, LEAF_ALGO>();
        }
      };
    };
  };
}

template <typename STORAGE_TYPE>
void infer_k_launcher(predict_params& params, storage_type forest,
                      cudaStream_t stream) {
  return template <bool COLS_IN_SHMEM>
  [&]() {
    return template <int N_ITEMS>
    [&]() {
      return template <leaf_algo_t LEAF_ALGO>
      [&]() {
        params.num_blocks =
          params.num_blocks != 0
            ? params.num_blocks
            : raft::ceildiv(int(params.num_rows), params.n_items);
        infer_k<n_items, leaf_algo, cols_in_shmem, storage_type>
          <<<params.num_blocks, params.blockdim_x, params.shm_sz, stream>>>(
            forest, params);
        CUDA_CHECK(cudaPeekAtLastError());
      };
    };
  };
}
```

Reply:
having tried this out, template lambdas are only supported in C++20. Simulating this with structs and `operator()` takes twice as many lines of wrapping (so +10-14 LOC/dispatchable), and eventually run into `error: invalid declaration of member template in local class`. So unless you have an idea, I don't think one can.